[view-transitions] Skip view transition when new element becomes hidden

nt1m opened the pull request:
#### 092ba08d7009635f9e0a789042e938a4c0cb47a9
<pre>
[view-transitions] Skip view transition when new element becomes hidden
<a href="https://bugs.webkit.org/show_bug.cgi?id=270815">https://bugs.webkit.org/show_bug.cgi?id=270815</a>
<a href="https://rdar.apple.com/124406569">rdar://124406569</a>

Reviewed by Ryosuke Niwa.

As mandated by the spec.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::captureOldState):
(WebCore::ViewTransition::activateViewTransition):
(WebCore::ViewTransition::updatePseudoElementStyles):
* Source/WebCore/dom/ViewTransition.h:

Canonical link: <a href="https://commits.webkit.org/275956@main">https://commits.webkit.org/275956@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69a4e0854469b5aad304c85f9c6ec835a929a069

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43335 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22365 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45745 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45970 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39461 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45639 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26115 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19783 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35826 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43909 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19387 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37329 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16806 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16969 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38387 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1391 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39519 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38706 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47514 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18249 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15016 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42612 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19786 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41270 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/9653 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19965 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5897 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19417 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/dom/ViewTransition.h:
```diff
@@ -157,7 +157,7 @@ class ViewTransition : public RefCounted<ViewTransition>, public CanMakeWeakPtr<
     Ref<MutableStyleProperties> copyElementBaseProperties(Element&, const LayoutSize&);
 
-    void updatePseudoElementStyles();
+    ExceptionOr<void> updatePseudoElementStyles();
```
Is it exposed in the JS API ?

nt1m replied:
The spec says to handle the exception here:
* https://www.w3.org/TR/css-view-transitions-1/#activate-view-transition
* https://www.w3.org/TR/css-view-transitions-1/#handle-transition-frame-algorithm

nt1m replied:
> [Update pseudo-element styles](https://www.w3.org/TR/css-view-transitions-1/#update-pseudo-element-styles) for transition.

> If failure is returned, then [skip the view transition](https://www.w3.org/TR/css-view-transitions-1/#skip-the-view-transition) for transition with an "[InvalidStateError](https://webidl.spec.whatwg.org/#invalidstateerror)" [DOMException](https://webidl.spec.whatwg.org/#idl-DOMException) in transition’s [relevant Realm](https://html.spec.whatwg.org/multipage/webappapis.html#concept-relevant-realm), and return.

mdubet replied:
> Return failure if any of the following conditions is true:
> ...
> capturedElement’s new element is not rendered

Seems like it

nt1m committed:
[view-transitions] Skip view transition when new element becomes hidden
https://bugs.webkit.org/show_bug.cgi?id=270815
rdar://124406569

Reviewed by Ryosuke Niwa.

As mandated by the spec.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::captureOldState):
(WebCore::ViewTransition::activateViewTransition):
(WebCore::ViewTransition::updatePseudoElementStyles):
* Source/WebCore/dom/ViewTransition.h:

Canonical link: https://commits.webkit.org/275956@main
